File: ios/DesMoinesInsider/Services/SessionTimeoutService.swift
import Foundation
import os

/// Tracks user activity and enforces session timeouts matching backend session_policies.
///
/// - Regular users: 60-min idle timeout, 8-hour absolute timeout
/// - Admin users: 30-min idle timeout, 4-hour absolute timeout
///
/// Resets idle timer on scene phase changes (active → background → active).
/// Persists last activity timestamp in Keychain for app restart scenarios.
@MainActor
@Observable
final class SessionTimeoutService {
    static let shared = SessionTimeoutService()

    // MARK: - Configuration

    private enum Timeout {
        // User timeouts
        static let userIdle: TimeInterval = 60 * 60           // 60 minutes
        static let userAbsolute: TimeInterval = 8 * 60 * 60   // 8 hours

        // Admin timeouts (stricter)
        static let adminIdle: TimeInterval = 30 * 60           // 30 minutes
        static let adminAbsolute: TimeInterval = 4 * 60 * 60   // 4 hours

        // Warning before expiry
        static let warningBefore: TimeInterval = 5 * 60        // 5 minutes

        // Check interval
        static let checkInterval: TimeInterval = 30            // 30 seconds
    }

    private enum Keys {
        static let lastActivity = "session_last_activity"
        static let sessionStart = "session_start_time"
        static let isAdmin = "session_is_admin"
    }

    // MARK: - State

    enum SessionState: Equatable {
        case active
        case warning(minutesRemaining: Int)
        case expired
    }

    private(set) var sessionState: SessionState = .active
    private var isAdmin = false
    private var isTracking = false
    private var checkTask: Task<Void, Never>?

    /// Last time `recordActivity` actually persisted, used to throttle Keychain
    /// writes when activity is driven by a continuous gesture (scroll/drag).
    private var lastRecordedActivity: TimeInterval = 0

    // MARK: - Lifecycle

    /// Start tracking session timeouts. Call after successful authentication.
    func startTracking(isAdmin: Bool) {
        self.isAdmin = isAdmin
        self.isTracking = true

        // Persist the admin flag so a cold-launch `isSessionValid()` check
        // applies the correct (stricter) admin thresholds even before auth has
        // re-resolved the role for this launch.
        KeychainService.shared.saveString(key: Keys.isAdmin, value: isAdmin ? "1" : "0")

        let now = Date().timeIntervalSince1970
        KeychainService.shared.saveString(key: Keys.sessionStart, value: String(now))
        lastRecordedActivity = 0   // force the first recordActivity to persist
        recordActivity()

        // Start periodic check
        checkTask?.cancel()
        checkTask = Task { [weak self] in
            while !Task.isCancelled {
                self?.checkTimeouts()
                try? await Task.sleep(nanoseconds: UInt64(Timeout.checkInterval * 1_000_000_000))
            }
        }

        AppLogger.auth.info("Session timeout tracking started (admin=\(isAdmin))")
    }

    /// Stop tracking session timeouts. Call on sign out.
    func stopTracking() {
        isTracking = false
        checkTask?.cancel()
        checkTask = nil
        sessionState = .active

        KeychainService.shared.delete(key: Keys.lastActivity)
        KeychainService.shared.delete(key: Keys.sessionStart)
        KeychainService.shared.delete(key: Keys.isAdmin)
        lastRecordedActivity = 0

        AppLogger.auth.info("Session timeout tracking stopped")
    }

    /// Record user activity (resets the idle timer). Safe to call on every user
    /// interaction — including continuous gestures — because the Keychain write
    /// is throttled; idle resolution is 30s so sub-throttle staleness is
    /// irrelevant. A pending warning is always cleared immediately, though, so
    /// the banner disappears the instant the user interacts.
    func recordActivity() {
        if case .warning = sessionState {
            sessionState = .active
        }

        let now = Date().timeIntervalSince1970
        guard now - lastRecordedActivity >= 5 else { return }
        lastRecordedActivity = now
        KeychainService.shared.saveString(key: Keys.lastActivity, value: String(now))
    }

    /// Check if the session is still valid on app restart.
    func isSessionValid() -> Bool {
        guard let lastActivityStr = KeychainService.shared.loadString(key: Keys.lastActivity),
              let sessionStartStr = KeychainService.shared.loadString(key: Keys.sessionStart),
              let lastActivity = Double(lastActivityStr),
              let sessionStart = Double(sessionStartStr) else {
            return true // No tracking data
        }

        // Use the persisted admin flag rather than the in-memory `isAdmin`, which
        // is still its `false` default this early in cold launch — otherwise an
        // admin would be granted the looser user thresholds.
        let wasAdmin = KeychainService.shared.loadString(key: Keys.isAdmin) == "1"

        let now = Date().timeIntervalSince1970
        let idleTimeout = wasAdmin ? Timeout.adminIdle : Timeout.userIdle
        let absoluteTimeout = wasAdmin ? Timeout.adminAbsolute : Timeout.userAbsolute

        let idleExpired = (now - lastActivity) > idleTimeout
        let absoluteExpired = (now - sessionStart) > absoluteTimeout

        return !idleExpired && !absoluteExpired
    }

    // MARK: - Private

    private func checkTimeouts() {
        guard isTracking else { return }

        guard let lastActivityStr = KeychainService.shared.loadString(key: Keys.lastActivity),
              let sessionStartStr = KeychainService.shared.loadString(key: Keys.sessionStart),
              let lastActivity = Double(lastActivityStr),
              let sessionStart = Double(sessionStartStr) else {
            return
        }

        let now = Date().timeIntervalSince1970
        let idleTimeout = isAdmin ? Timeout.adminIdle : Timeout.userIdle
        let absoluteTimeout = isAdmin ? Timeout.adminAbsolute : Timeout.userAbsolute

        let idleElapsed = now - lastActivity
        let absoluteElapsed = now - sessionStart

        // Check absolute timeout first (non-resettable)
        if absoluteElapsed >= absoluteTimeout {
            sessionState = .expired
            AppLogger.auth.warning("Session expired: absolute timeout reached")
            return
        }

        // Check idle timeout
        if idleElapsed >= idleTimeout {
            sessionState = .expired
            AppLogger.auth.warning("Session expired: idle timeout reached")
            return
        }

        // Check warning threshold
        let idleRemaining = idleTimeout - idleElapsed
        let absoluteRemaining = absoluteTimeout - absoluteElapsed
        let minRemaining = min(idleRemaining, absoluteRemaining)

        if minRemaining <= Timeout.warningBefore {
            let minutes = max(Int(minRemaining / 60), 1)
            sessionState = .warning(minutesRemaining: minutes)
        } else {
            sessionState = .active
        }
    }

    private init() {}
}

File: android/app/src/main/java/com/desmoines/aipulse/util/SessionTimeoutService.kt
package com.desmoines.aipulse.util

import android.app.Activity
import android.app.Application
import android.os.Bundle
import dagger.hilt.android.qualifiers.ApplicationContext
import android.content.Context
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Tracks admin activity and enforces session timeouts: 30 minutes idle,
 * 4 hours absolute.
 *
 * ADMIN SESSIONS ONLY. The backend `session_policies` table also defines a
 * 60-minute / 8-hour policy for regular users, and this class used to carry it
 * too. The client deliberately does not enforce that half (AND-AUDIT-006): this
 * is a consumer events and restaurant guide, and signing a browsing user out
 * mid-listing is a real cost against no threat model. A stale *admin* session is
 * where it actually matters. Callers gate on the admin role; [startTracking] is
 * only ever called for one.
 *
 * The idle timer is reset by [recordActivity], which callers fire on navigation
 * and on return to the foreground. Do not rely on the lifecycle callback below
 * alone: the app is single-Activity Compose, so `onActivityResumed` fires when
 * the app is re-entered and never while someone is using it. Wired to that
 * signal by itself, this class would sign an admin out mid-scroll.
 *
 * Last-activity and session-start timestamps persist in SecureStorage so
 * [isSessionValid] can catch a session that expired while the app was closed.
 */
@Singleton
class SessionTimeoutService @Inject constructor(
    @param:ApplicationContext private val context: Context,
    private val secureStorage: SecureStorage,
) {

    companion object {
        private const val KEY_LAST_ACTIVITY = "session_last_activity"
        private const val KEY_SESSION_START = "session_start_time"

        private const val IDLE_TIMEOUT_MS = 30L * 60 * 1000        // 30 minutes
        private const val ABSOLUTE_TIMEOUT_MS = 4L * 60 * 60 * 1000 // 4 hours

        // Warning before expiry
        private const val WARNING_BEFORE_MS = 5L * 60 * 1000           // 5 minutes

        // Check interval
        private const val CHECK_INTERVAL_MS = 30_000L                  // 30 seconds
    }

    // Reads and writes EncryptedSharedPreferences on every tick; the state it
    // publishes is a StateFlow, which is safe to update from any thread.
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private var checkJob: Job? = null

    // Named isTracking, not isActive. Inside scope.launch { } the lambda
    // receiver is CoroutineScope, which carries its own isActive extension
    // property, and that shadows a field of the same name -- so the loop below
    // read the coroutine's liveness and this flag was never actually consulted.
    @Volatile
    private var isTracking = false

    private val _sessionState = MutableStateFlow(SessionState.ACTIVE)
    val sessionState: StateFlow<SessionState> = _sessionState.asStateFlow()

    private val _minutesRemaining = MutableStateFlow<Int?>(null)
    val minutesRemaining: StateFlow<Int?> = _minutesRemaining.asStateFlow()

    enum class SessionState {
        ACTIVE,
        WARNING,
        EXPIRED,
    }

    private val activityCallbacks = object : Application.ActivityLifecycleCallbacks {
        override fun onActivityResumed(activity: Activity) {
            recordActivity()
        }
        override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {}
        override fun onActivityStarted(activity: Activity) {}
        override fun onActivityPaused(activity: Activity) {}
        override fun onActivityStopped(activity: Activity) {}
        override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {}
        override fun onActivityDestroyed(activity: Activity) {}
    }

    /**
     * Start tracking. Call after an authenticated user is confirmed to be an
     * admin; see the class doc for why non-admin sessions are not tracked.
     *
     * Resets the idle timer, so check [isSessionValid] first if you need to know
     * whether the previous session had already expired.
     */
    fun startTracking() {
        isTracking = true

        val now = System.currentTimeMillis()
        secureStorage.saveLong(KEY_SESSION_START, now)
        recordActivity()

        // Register lifecycle callbacks
        (context.applicationContext as? Application)?.registerActivityLifecycleCallbacks(activityCallbacks)

        // Start periodic timeout checks
        checkJob?.cancel()
        checkJob = scope.launch {
            while (isTracking) {
                checkTimeouts()
                delay(CHECK_INTERVAL_MS)
            }
        }

        AppLogger.auth.info("Admin session timeout tracking started")
    }

    /**
     * Stop tracking session timeouts. Call on sign out.
     */
    fun stopTracking() {
        isTracking = false
        checkJob?.cancel()
        checkJob = null
        _sessionState.value = SessionState.ACTIVE
        _minutesRemaining.value = null

        secureStorage.delete(KEY_LAST_ACTIVITY)
        secureStorage.delete(KEY_SESSION_START)

        (context.applicationContext as? Application)?.unregisterActivityLifecycleCallbacks(activityCallbacks)

        AppLogger.auth.info("Session timeout tracking stopped")
    }

    /**
     * Record user activity (resets idle timer).
     */
    fun recordActivity() {
        secureStorage.saveLong(KEY_LAST_ACTIVITY, System.currentTimeMillis())
        if (_sessionState.value == SessionState.WARNING) {
            _sessionState.value = SessionState.ACTIVE
            _minutesRemaining.value = null
        }
    }

    /**
     * Check if the session has timed out on app restart.
     * Returns true if the session is still valid.
     */
    fun isSessionValid(): Boolean {
        val lastActivity = secureStorage.loadLong(KEY_LAST_ACTIVITY)
        val sessionStart = secureStorage.loadLong(KEY_SESSION_START)

        if (lastActivity == 0L || sessionStart == 0L) return true // No tracking data

        val now = System.currentTimeMillis()
        val idleExpired = (now - lastActivity) > IDLE_TIMEOUT_MS
        val absoluteExpired = (now - sessionStart) > ABSOLUTE_TIMEOUT_MS

        return !idleExpired && !absoluteExpired
    }

    private fun checkTimeouts() {
        val lastActivity = secureStorage.loadLong(KEY_LAST_ACTIVITY)
        val sessionStart = secureStorage.loadLong(KEY_SESSION_START)
        val now = System.currentTimeMillis()

        if (lastActivity == 0L || sessionStart == 0L) return

        val idleElapsed = now - lastActivity
        val absoluteElapsed = now - sessionStart

        // Check absolute timeout first (non-resettable)
        if (absoluteElapsed >= ABSOLUTE_TIMEOUT_MS) {
            _sessionState.value = SessionState.EXPIRED
            AppLogger.auth.warning("Session expired: absolute timeout reached")
            return
        }

        // Check idle timeout
        if (idleElapsed >= IDLE_TIMEOUT_MS) {
            _sessionState.value = SessionState.EXPIRED
            AppLogger.auth.warning("Session expired: idle timeout reached")
            return
        }

        // Check warning threshold
        val idleRemaining = IDLE_TIMEOUT_MS - idleElapsed
        val absoluteRemaining = ABSOLUTE_TIMEOUT_MS - absoluteElapsed
        val minRemaining = minOf(idleRemaining, absoluteRemaining)

        if (minRemaining <= WARNING_BEFORE_MS) {
            _sessionState.value = SessionState.WARNING
            _minutesRemaining.value = (minRemaining / 60_000).toInt().coerceAtLeast(1)
        } else {
            _sessionState.value = SessionState.ACTIVE
            _minutesRemaining.value = null
        }
    }
}
